lib/src/spec/message.dart

```diff
@@ -2,16 +2,40 @@ import 'enums.dart';
 import 'rest/channels.dart';
 
 class Message {
-  Message.fromEncoded(Map jsonObject, ChannelOptions channelOptions);
-  Message.fromEncodedArray(List<Map> jsonArray, ChannelOptions channelOptions);
+
+  /// A unique ID for this message
+  String id;
+
+  /// The timestamp for this message
+  DateTime timestamp;
+
+  /// The id of the publisher of this message
   String clientId;
+
+  /// The connection id of the publisher of this message
   String connectionId;
-  dynamic data;
+
+  /// Any transformation applied to the data for this message
   String encoding;
-  dynamic extras;
-  String id;
+
+  /// The message payload
+  dynamic data;
```

Comment:
I think `dynamic` is almost always a bad idea (except when implementing `noSuchMethod`).
Either use `Object` or any more concrete class (or interface) if possible.
@QuintinWillison what do you think?

Reply:
I think object should do.
Acceptable types for `data` are `null`, `String`, `Map` and `List`

Reply:
ok, in this case object (and dynamic) are the only options

Reply:
addressed 9b9d8a44b61653d442cd5b90c49481fbb67b78ff

Reply:
Thanks both. My preference will always be strongly typed but, alas, we're somewhat beholden to [the spec](https://docs.ably.io/client-lib-development-guide/features/#TM2d).

There could be an opportunity to do something more strongly typed here so I would be open to suggestions. Perhaps `data` could present an API that offers the app developer to use language akin to the `asString()`, `asBytes()` seen elsewhere?

Feels like it cuts to the core of Ably's "conformed APIs" approach (vs idiomatic) so would be interested in views from @paddybyers, @mattheworiordan and @SimonWoolf.

Reply:
@QuintinWillison _Just FYI_, in `ably-java` we have **message payload** as `Object` and in `ably-cocoa` as `id`

Reply:
I have the same opinion about strongly typed, but it looks there are no good options that would be a real improvement.
A wrapper class could be used with a field per property value type, but `Map` and `List` would need generic parameters to actually provide notably more type safety than Object.

In the message receive case that's hardly possible because I guess we usually do not know what type the message will have at compile time.

Reply:
We don't but, in my opinion, the app developer might reasonably know - therefore they then call the correct helper method (or is that getter) on the returned object.

Reply:
We could wait until Dart supports union types which seems to be part of the next bigger chunk on the roadmap https://github.com/dart-lang/language/issues/546 which might take a few years 
or create a wrapper class similar to a union type ourselves until Dart provides that.

Reply:
I'd suggest to create a new issue for that instead of making in part of this PR

Reply:
Personally I think we should go with what's easiest to implement now and avoid introducing new "special" types that deviate significantly from other SDKs.

Reply:
@QuintinWillison Sorry, I didn't mean to dismiss your suggestion with different accessor methods/getters. 
It's rather that I can't imagine how that would work out, but that's definitely because I do not know Ably's API well yet.

Reply:
Created an issue to move this discussion forward: https://github.com/ably/ably-flutter/issues/35

I've also added some inputs based on @QuintinWillison's suggestion